.zsh/prompt.zsh

```diff
@@ -38,32 +27,29 @@ ZSH_HIGHLIGHT_STYLES[comment]='fg=white,bold'
   typeset -g POWERLEVEL9K_{LEFT,RIGHT}_SUBSEGMENT_SEPARATOR=' '
   typeset -g POWERLEVEL9K_{LEFT,RIGHT}_SEGMENT_SEPARATOR=
 
-  typeset -g POWERLEVEL9K_VISUAL_IDENTIFIER_EXPANSION='${P9K_VISUAL_IDENTIFIER// }'
   typeset -g POWERLEVEL9K_MODE=awesome-fontconfig
   typeset -g POWERLEVEL9K_PROMPT_ADD_NEWLINE=true
-  typeset -g POWERLEVEL9K_MULTILINE_FIRST_PROMPT_GAP_CHAR=' '
 
   ################################[ background_jobs: background jobs indicator ]################
   typeset -g POWERLEVEL9K_BACKGROUND_JOBS_FOREGROUND=red
   typeset -g POWERLEVEL9K_BACKGROUND_JOBS_VISUAL_IDENTIFIER_EXPANSION=" "
 
-  ################################[ context: shows root user ]#################################
-  typeset -g DEFAULT_USER=maximbaz
-  typeset -g POWERLEVEL9K_CONTEXT_FOREGROUND=red
-  typeset -g POWERLEVEL9K_CONTEXT_TEMPLATE="%n"
-  typeset -g POWERLEVEL9K_CONTEXT_PREFIX="%fas "
+  ################################[ user: shows root user ]#################################
+  typeset -g POWERLEVEL9K_USER_FOREGROUND=red
+  typeset -g POWERLEVEL9K_USER_PREFIX="%fas "
+  typeset -g POWERLEVEL9K_USER_CONTENT_EXPANSION=
+  typeset -g POWERLEVEL9K_USER_ROOT_CONTENT_EXPANSION='${P9K_CONTENT}'
```

Comment:
Oh nice, this works great, I only need to add `typeset -g POWERLEVEL9K_ROOT_ICON=''` otherwise I'm getting a weird icon in front:

![image](https://user-images.githubusercontent.com/1177900/66276908-bb348d80-e897-11e9-8792-6b8ef6f704e6.png)

Unless this is not the best way?

Reply:
Oops, missed that.

You are using so few icons that it's easier to disable them by default (the same way background is disabled by default) and then explicitly enabled the ones you want. I've force-pushed this change now.

Reply:
Fair enough 😄

Reply:
In case you decide to further customize your prompt, consider writing custom prompt segments. It's easy, and they are fast (provided you don't work in them). For example, your `user` segment can be replaced with `root` that is defined as follows:

```zsh
function prompt_root() {
  # Red $USER if running with privileges, otherwise nothing.
  (( EUID )) || p10k segment -f red -t '%n'
}
```

That's all it takes. You can use `root` segment as if it was builtin. All options such as `POWERLEVEL9K_ROOT_FOREGROUND` will work.

Type `p10k help segment` for details.

Reply:
This is cool! Is there a way to make the segment load asynchronously? 

I have one particularly slow command that I need to put in the prompt, with spaceship it wasn't an issue because it will show up in the prompt whenever loading finishes, but I can't find how to achieve the same in p10k...

Let's say this is the segment, is it possible to have it and yet not make the entire prompt hang while waiting for it to calculate?

```
  function prompt_mysegment() {
    local val=$(sleep 3; echo mysegment)
    p10k segment -f red -t "$val"
  }
```

Reply:
(It's more of a theoretical question, as I've just figured out how to drastically improve the performance of that command...)

Reply:
> (It's more of a theoretical question, as I've just figured out how to drastically improve the performance of that command...)

👍 

Can you share the source of this segment? I'm always on lookout for new segments to add to p10k.

p10k doesn't expose an API for defining async segments even though there are two internal async segment in it. These segments are `vcs` and `public_ip`. `vcs` has async support because its latency is `O(N)` where N is the number of files in the Git index and it's unbound. In practice you usually start seeing `vcs` in `LOADING` state when your Git repo has somewhere between 100k and 1M files. `public_ip` is async because it performs network IO to do its job, which can take seconds.

If you really need to, you can implement an async segment for p10k. It's not much code and doesn't require changes to p10k proper. You have to make several decisions that cannot be made for you by p10k: when to start the async command (on every prompt? every 5 seconds? when some file changes? when the current directory changes?); what to display while the async command is running (nothing? previous state but greyed out? previous state from the same directory? something else?) These questions make it difficult to implement generic API for async segments that doesn't limit their flexibility. Unless your segment communicates over the network or needs to rip through tens of thousands of files, it's likely possible to implement it efficiently without going async. It'll be better UX.

Reply:
> Can you share the source of this segment? I'm always on lookout for new segments to add to p10k.

Here it is https://github.com/maximbaz/dotfiles/commit/e122639b3183714a91fae1b4906f5f7d367c47e5, it's a segment that shows current Azure subscription. I used to use `az account show --query name --output tsv 2>/dev/null` in spaceship but this command is crazy slow, and I've been testing for the past 30 mins, the `jq` alternative I came up with seems to work well (and fast).

Makes sense about the rest, thanks for explaining and leaving some pointers!